Update button activation and reflection behavior with command or commandfor

lukewarlow opened the pull request:
#### b4d27c3a21dab43e99a8c79d24a59aed7c0761e4
<pre>
Update button activation and reflection behavior with command or commandfor
<a href="https://bugs.webkit.org/show_bug.cgi?id=287288">https://bugs.webkit.org/show_bug.cgi?id=287288</a>

Reviewed by Anne van Kesteren.

This patch correctly updates the button activation steps to match the spec,
missing and invalid type buttons no longer submit forms when they have a command or commandfor
attribute. They also correctly don&apos;t act as command buttons.

This patch also updates form associated reset buttons and inputs to not trigger
popovertarget or commandfor behavior.

The button elements type reflection also correctly reflects as button
when it has a missing or invalid type attribute and has command or commandfor.

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/button-type-popovertarget-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/input-type-popovertarget-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/the-button-element/command-and-commandfor/button-type-behavior-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/the-button-element/command-and-commandfor/button-type-reflection-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/html/semantics/the-button-element/command-and-commandfor/button-type-behavior-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/html/semantics/the-button-element/command-and-commandfor/button-type-reflection-expected.txt:
* Source/WebCore/html/HTMLButtonElement.cpp:
(WebCore::HTMLButtonElement::attributeChanged):
(WebCore::HTMLButtonElement::defaultEventHandler):
(WebCore::HTMLButtonElement::computeType):
* Source/WebCore/html/HTMLButtonElement.h:
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::defaultEventHandler):

Canonical link: <a href="https://commits.webkit.org/292010@main">https://commits.webkit.org/292010@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f5e8c91c6a83fd9527f651fbde89ae5ba67c4d5e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/94515 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/14107 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/3923 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/99534 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/45031 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/96565 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/14405 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/22535 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/72121 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/29445 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/97517 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/10722 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/85429 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/52453 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/10415 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/3054 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/44354 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/80635 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/3160 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/101578 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/21571 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/15735 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/81124 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/21821 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/81375 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/80500 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20122 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/25053 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/2450 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/14794 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/21548 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/26683 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/115829 "Build is in progress. Recent messages:") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/21227 "Built successfully") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/35/builds/115829 "Build is in progress. Recent messages:") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/24690 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/22965 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk reviewed (commented):
Hmm, wouldn't it be better if we changed how m_type is computed?

annevk commented on Source/WebCore/html/HTMLButtonElement.cpp:
```diff
@@ -342,4 +337,25 @@ bool HTMLButtonElement::isExplicitlySetSubmitButton() const
 
+void HTMLButtonElement::computeType(const AtomString& typeAttrValue)
+{
+    Type oldType = m_type;
```
```suggestion
    auto oldType = m_type;
```

annevk commented on Source/WebCore/html/HTMLButtonElement.cpp:
```diff
@@ -342,4 +337,25 @@ bool HTMLButtonElement::isExplicitlySetSubmitButton() const
+        m_type = RESET;
+    else if (equalLettersIgnoringASCIICase(typeAttrValue, "button"_s))
+        m_type = BUTTON;
+    else if (typeAttrValue.isEmpty() && document().settings().invokerAttributesEnabled()) {
```
This seems wrong. Shouldn't this check that it's not the string `submit`?

lukewarlow replied:
It was correct at the time it was written but yes it now needs changing. Will mark as draft until I've resynced the tests and updated to match the spec.

nt1m commented on Source/WebCore/html/HTMLInputElement.cpp:
```diff
@@ -1357,6 +1357,9 @@ void HTMLInputElement::defaultEventHandler(Event& event)
     if (event.type() == eventNames().DOMActivateEvent) {
         m_inputType->handleDOMActivateEvent(event);
+        RefPtr<HTMLFormElement> protectedForm(form());
+        if (protectedForm && m_inputType->type() != InputType::Type::Button)
```
Protectors are generally outdated patterns for smart pointers, also they're generally not needed for null checks.

```suggestion
        if (form() && m_inputType->type() != InputType::Type::Button)
```

nt1m commented on Source/WebCore/html/HTMLButtonElement.cpp:
```diff
@@ -225,13 +216,7 @@ void HTMLButtonElement::defaultEventHandler(Event& event)
-            handleCommand();
-
-        } else if (protectedForm) {
+        if (protectedForm) {
```
Same comment about the protector

nt1m commented on Source/WebCore/html/HTMLButtonElement.cpp:
```diff
@@ -244,13 +229,21 @@ void HTMLButtonElement::defaultEventHandler(Event& event)
                     currentForm->reset();
```
I notice there is a null check for currentForm which is duplicated with protectedForm...

It should probably just be:

```
Ref currentForm = *form();
```

without the `if ()`

nt1m reviewed (commented):
The flow of `HTMLButtonElement::defaultEventHandler` has gone slightly confusing, perhaps there's some cleanup to be done before adding more code.

annevk commented on Source/WebCore/html/HTMLButtonElement.cpp:
```diff
@@ -223,34 +214,33 @@ void HTMLButtonElement::defaultEventHandler(Event& event)
-            if (RefPtr currentForm = form()) {
-                if (m_type == SUBMIT)
-                    currentForm->submitIfPossible(&event, this);
+            RefPtr currentForm = *form()
```
I think the previous code was correct if layout can indeed cause form to become null.

(This seems anyway incorrect. If it's always non-null it should be Ref and there should be a semi-colon at the end.

lukewarlow replied:
Can layout cause form to be null? I would think no but perhaps I'm missing something?

Good catch on the missing semi colon I was trying to pre-empty my local build but clearly messed up.

annevk replied:
The comment above and the existing null check seem to suggest it can. I'm also a bit baffled by it, but presumably that's there for some reason. Don't have time to investigate right now though.

lukewarlow replied:
I'll play it safe and assume it does so I'll restore the if check

annevk commented on Source/WebCore/html/HTMLButtonElement.cpp:
```diff
@@ -342,4 +332,27 @@ bool HTMLButtonElement::isExplicitlySetSubmitButton() const
+        m_type = SUBMIT;
+    if (oldType != m_type) {
+        updateWillValidateAndValidity();
+        if (form() && (oldType == SUBMIT || m_type == SUBMIT))
```
Here you probably do want to RefPtr form given you call a method on it.

lukewarlow committed:
Update button activation and reflection behavior with command or commandfor
https://bugs.webkit.org/show_bug.cgi?id=287288

Reviewed by Anne van Kesteren.

This patch correctly updates the button activation steps to match the spec,
missing and invalid type buttons no longer submit forms when they have a command or commandfor
attribute. They also correctly don't act as command buttons.

This patch also updates form associated reset buttons and inputs to not trigger
popovertarget or commandfor behavior.

The button elements type reflection also correctly reflects as button
when it has a missing or invalid type attribute and has command or commandfor.

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/button-type-popovertarget-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/input-type-popovertarget-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/the-button-element/command-and-commandfor/button-type-behavior-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/the-button-element/command-and-commandfor/button-type-reflection-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/html/semantics/the-button-element/command-and-commandfor/button-type-behavior-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/html/semantics/the-button-element/command-and-commandfor/button-type-reflection-expected.txt:
* Source/WebCore/html/HTMLButtonElement.cpp:
(WebCore::HTMLButtonElement::attributeChanged):
(WebCore::HTMLButtonElement::defaultEventHandler):
(WebCore::HTMLButtonElement::computeType):
* Source/WebCore/html/HTMLButtonElement.h:
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::defaultEventHandler):

Canonical link: https://commits.webkit.org/292010@main
